docs/device-assignment.md

```diff
@@ -0,0 +1,130 @@
+
+The VFIO endpoints, for devices requested by the pod are then exposed as host/container paths (/dev/vfio/vfio, /dev/vfio/{N, M, ...} where N, M are IOMMU group IDs) to the container in the pod.
+
+The main difficulty of this approach is reliably identifying which system device was assigned. The KubeVirt API must be verbose enough to let users request a device by it's vendor ID/product ID tuple, yet hide the actual device address. This is difficult for libvirt, as VM XML requires users to specify the device address.
+
+Because the actual device selection is done by kube-scheduler, and system may contain multiple devices with identical vendor ID/product ID that are in the same or different IOMMU groups, we need a mechanism to inspect which IOMMU group does the device belong to.
+
+To demonstrate the issue, we present an example.
+
+### Example
+
+* IOMMU group `1` contains devices
+ *	`A` (vendor ID = 1000, product ID = 1000)
+ * `B` (vendor ID = 2000, product ID = 2000)
+* IOMMU group `2` contains
+ * device `C` (vendor ID = 1000, product ID = 1000)
+* we have VM `α`, and it's corresponding pod `a`
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachine
+metadata:
+  name: α
+spec:
+  domain:
+    devices:
+      ...
+      passthrough:
+      - type: pci
+        vendor: 1000
+        device: 1000
+      ...
+    memory:
+      unit: MB
+      value: 64
+    os:
+      type:
+        os: hvm
+    type: qemu
+```
+
+In this case, KubeVirt must generate pod specification that requests device with vendor ID/product ID 1000:1000. Such specification is shown below.
+
+```yaml
+spec:
+  containers:
+  - name: a
+    ...
+    resources:
+      requests:
+              mpolednik.github.io/1000_1000: 1
+      limits:
+              mpolednik.github.io/1000_1000: 1
+```
+
+The difficult part is: how should libvirt XML generated by virt-handler look like? It is able to query the underlying system for devices and determine addresses of all devices that match vendor ID/product ID 1000:1000, but choosing a different device than the scheduler scheduled would cause the VM not to start due to IOMMU violation.
+
+To safely identify the selected device, we have to extend `virt-launcher` to report back the IOMMU groups inside the VMs pod. This could be done via a file in a directory shared between virt-launcher and virt-handler. Knowing which IOMMU groups were assigned to the pod, virt-handler can then probe the system and find device addresses to construct the final libvirt XML.
```

Comment:
By report-back, you mean reporting it to libvirt, right?

Reply:
My idea is more of

1. virt-launcher probes /dev/vfio/*, communicates that to virt-handler via file in the shared dir,
2. virt-handler (either directly or through an extension) reads device IDs and addresses from the corresponding IOMMU group in sysfs (/sys/kernel/iommu_groups/groupN/devices),
3. virt-handler sets the device address in VM XML.

Libvirt isn't really suited for the cloudy version of device assignment, so we have to help it a bit.